molgenis-data-platform/src/main/java/org/molgenis/data/platform/decorators/AuditingRepositoryDecoratorFactory.java

```diff
@@ -0,0 +1,97 @@
+package org.molgenis.data.platform.decorators;
+
+import static com.google.common.collect.Streams.stream;
+import static java.util.Objects.requireNonNull;
+import static org.molgenis.data.index.job.IndexJobExecutionMetadata.INDEX_JOB_EXECUTION;
+import static org.molgenis.data.index.meta.IndexActionGroupMetadata.INDEX_ACTION_GROUP;
+import static org.molgenis.data.index.meta.IndexActionMetadata.INDEX_ACTION;
+import static org.molgenis.data.semantic.Relation.isAudited;
+import static org.molgenis.data.semantic.Vocabulary.AUDIT_USAGE;
+import static org.molgenis.data.util.EntityTypeUtils.isSystemEntity;
+import static org.molgenis.security.audit.AuditSettingsImpl.AUDIT_SETTINGS;
+
+import com.google.common.collect.ImmutableSet;
+import org.molgenis.audit.AuditEventPublisher;
+import org.molgenis.data.Entity;
+import org.molgenis.data.Repository;
+import org.molgenis.data.meta.model.EntityType;
+import org.molgenis.data.meta.model.Tag;
+import org.molgenis.data.security.audit.AuditingRepositoryDecorator;
+import org.molgenis.security.audit.AuditSettings;
+import org.molgenis.util.UnexpectedEnumException;
+import org.springframework.stereotype.Component;
+
+/**
+ * Decorates a {@link Repository} with an {@link AuditingRepositoryDecorator} if the Repository
+ * meets the following criteria:
+ *
+ * <p>For system entity types:
+ *
+ * <ul>
+ *   <li>Auditing of system entity types is enabled in the {@link AuditSettings}
+ *   <li>The repository is not excluded from auditing
+ * </ul>
+ *
+ * For data entity types:
+ *
+ * <ul>
+ *   <li>Auditing of data entity types is set to ALL in the {@link AuditSettings}, OR
+ *   <li>Auditing of data entity types is set to TAGGED in the {@link AuditSettings} and the entity
+ *       type is tagged with the 'audit-usage' tag.
+ * </ul>
```

Comment:
I'd put the evaluation of these criteria in a method, that's actually easier to read then